lightning/src/routing/router.rs

```diff
@@ -917,7 +917,7 @@ where L::Target: Logger {
 					// Since we're choosing amount_to_transfer_over_msat as maximum possible, it can
 					// be only reduced later (not increased), so this channel should just be skipped
 					// as not sufficient.
-					if !over_path_minimum_msat && doesnt_exceed_cltv_delta_limit {
+					if contributes_sufficient_value && !over_path_minimum_msat && doesnt_exceed_cltv_delta_limit {
```

Comment:
Oh, this isn't quite sufficient, either, lets *also* add a `&& recommended_value_msat <= both of the two comparisons for `over_path_minimum_msat`

Reply:
As in, don't bother setting `hit_minimum_limit` unless we both were not under the path minimum msat, and also the next amount we'll try (the `recommended_value_msat` amount) *is* below the path minimum msat (and channel minimum msat).

Reply:
Specifically, the issue here, I believe, is that some people are disabling channels by setting the minimum to 21 million bitcoin, and as a result we're always setting `hit_minimum_limit` and going around the dijkstras again, which is nonsense.

Reply:
Is this what you mean? Seeing a lot of failing tests.

```diff
diff --git a/lightning/src/routing/router.rs b/lightning/src/routing/router.rs
index 5a4ddb26..4d79cf63 100644
--- a/lightning/src/routing/router.rs
+++ b/lightning/src/routing/router.rs
@@ -913,8 +913,11 @@ where L::Target: Logger {
                                                None => unreachable!(),
                                        };
                                        #[allow(unused_comparisons)] // $next_hops_path_htlc_minimum_msat is 0 in some calls so rustc complains
-                                       let over_path_minimum_msat = amount_to_transfer_over_msat >= $candidate.htlc_minimum_msat() &&
-                                               amount_to_transfer_over_msat >= $next_hops_path_htlc_minimum_msat;
+                                       let over_path_minimum_msat =
+                                               amount_to_transfer_over_msat >= $candidate.htlc_minimum_msat() &&
+                                               amount_to_transfer_over_msat >= $next_hops_path_htlc_minimum_msat &&
+                                               recommended_value_msat <= $candidate.htlc_minimum_msat() &&
+                                               recommended_value_msat <= $next_hops_path_htlc_minimum_msat;
 
                                        // If HTLC minimum is larger than the amount we're going to transfer, we shouldn't
                                        // bother considering this channel.

```

Reply:
Okay, its a bittt more subtle for a few reasons, but this works:
```
let over_path_minimum_msat = amount_to_transfer_over_msat >= $candidate.htlc_minimum_msat() &&
    amount_to_transfer_over_msat >= $next_hops_path_htlc_minimum_msat;
    
#[allow(unused_comparisons)] // $next_hops_path_htlc_minimum_msat is 0 in some calls so rustc complains
let may_overpay_to_meet_path_minimum_msat = ((amount_to_transfer_over_msat < $candidate.htlc_minimum_msat() &&
        recommended_value_msat > $candidate.htlc_minimum_msat()) ||
    (amount_to_transfer_over_msat < $next_hops_path_htlc_minimum_msat &&
        recommended_value_msat > $next_hops_path_htlc_minimum_msat));
.....
                    if contributes_sufficient_value && may_overpay_to_meet_path_minimum_msat && doesnt_exceed_cltv_delta_limit {
                        hit_minimum_limit = true;
                    } else if contributes_sufficient_value && doesnt_exceed_cltv_delta_limit && over_path_minimum_msat {
```

In english: (a) we need to add an `over_path_minimum_msat` comparison on the second if, as we shouldn't use a channel at all if its not over the path min, even if we don't set `hit_minimum_limit`, (b) the new checks need to apply on the equivalent path-specific comparison - if we fail to meet the path htlc min, then recommended_value_msat has to be over the path htlc min (but not also the candidate min), if we fail to meet the hop htlc min, then the recommended value has to be over the hop htlc min (not the path htlc min).

I added a new variable above to make it a bit more readable, but its not clear to me that its the most readable version of the logic, feel free to swap it around some if you see a better way.

Reply:
I just swapped the ordering within the first if-condition so that it is symmetric with the second.